File: unreal/NeshamaSDK/Source/NeshamaSDK/Private/SoulEngine/Memory/MemorySystem.cpp
#include "SoulEngine/Memory/MemorySystem.h"

UMemorySystem::UMemorySystem()
{
}

const TMap<ESoulEventType, UMemorySystem::FRelationMapping>& UMemorySystem::GetEventRelationMappings()
{
	static TMap<ESoulEventType, FRelationMapping> Mappings;
	if (Mappings.Num() > 0) return Mappings;

	Mappings.Add(ESoulEventType::PlayerAttacked,  {TEXT("hostile"),       0.3f, -0.2f});
	Mappings.Add(ESoulEventType::PlayerHelped,     {TEXT("ally"),          0.3f,  0.2f});
	Mappings.Add(ESoulEventType::ItemReceived,     {TEXT("friendly"),      0.1f,  0.1f});
	Mappings.Add(ESoulEventType::ItemLost,         {TEXT("neutral"),      -0.1f, -0.1f});
	Mappings.Add(ESoulEventType::QuestCompleted,   {TEXT("ally"),          0.3f,  0.3f});
	Mappings.Add(ESoulEventType::QuestFailed,      {TEXT("disappointed"), -0.2f, -0.1f});
	Mappings.Add(ESoulEventType::NpcInsulted,      {TEXT("hostile"),       0.2f, -0.3f});
	Mappings.Add(ESoulEventType::NpcComplimented,  {TEXT("friendly"),      0.2f,  0.2f});
	Mappings.Add(ESoulEventType::GiftGiven,        {TEXT("friendly"),      0.2f,  0.3f});
	Mappings.Add(ESoulEventType::EnvironmentChanged,{TEXT("aware"),        0.1f,  0.0f});
	Mappings.Add(ESoulEventType::RelationshipChanged,{TEXT("connected"),   0.2f,  0.0f});
	Mappings.Add(ESoulEventType::CombatStarted,    {TEXT("hostile"),       0.2f,  0.0f});
	Mappings.Add(ESoulEventType::CombatEnded,      {TEXT("tense"),         0.0f,  0.0f});
	Mappings.Add(ESoulEventType::DeathWitnessed,   {TEXT("shaken"),        0.2f,  0.0f});
	Mappings.Add(ESoulEventType::TimePassed,       {TEXT("neutral"),      -0.05f, 0.0f});

	return Mappings;
}

void UMemorySystem::OnGameEvent(ESoulEventType EventType, float Intensity,
	const FString& EntityId, const FString& EntityName,
	const TMap<FString, float>& EmotionalContext)
{
	const auto& Mappings = GetEventRelationMappings();
	const FRelationMapping* MappingPtr = Mappings.Find(EventType);
	if (!MappingPtr) return;

	const FRelationMapping& Mapping = *MappingPtr;

	// Get or create relation
	FEntityRelation* RelationPtr = Relations.Find(EntityId);
	if (!RelationPtr)
	{
		FEntityRelation NewRelation;
		NewRelation.EntityId = EntityId;
		NewRelation.EntityName = EntityName;
		NewRelation.Strength = FMath::Clamp(0.3f + Mapping.StrengthDelta * Intensity, 0.0f, 1.0f);
		NewRelation.Trust = FMath::Clamp(0.3f + Mapping.TrustDelta * Intensity, 0.0f, 1.0f);
		NewRelation.RelationType = Mapping.Relation;
		NewRelation.LastInteractionTime = GameTime;
		NewRelation.InteractionCount = 1;
		Relations.Add(EntityId, NewRelation);
	}
	else
	{
		RelationPtr->Strength = FMath::Clamp(RelationPtr->Strength + Mapping.StrengthDelta * Intensity, -1.0f, 1.0f);
		RelationPtr->Trust = FMath::Clamp(RelationPtr->Trust + Mapping.TrustDelta * Intensity, 0.0f, 1.0f);
		RelationPtr->RelationType = Mapping.Relation;
		RelationPtr->LastInteractionTime = GameTime;
		RelationPtr->InteractionCount++;
	}

	// Create memory entry
	MemoryCounter++;
	FEntityMemory Memory;
	Memory.MemoryId = FString::Printf(TEXT("mem_%d"), MemoryCounter);
	Memory.EntityId = EntityId;
	Memory.EntityName = EntityName;
	Memory.EventType = StaticEnum<ESoulEventType>()->GetNameStringByValue(static_cast<int64>(EventType));
	Memory.Description = GenerateMemoryDescription(EventType, EntityName);
	Memory.Timestamp = GameTime;
	Memory.EmotionalContext = EmotionalContext;
	Memory.TrustAtTime = Relations.FindChecked(EntityId).Trust;
	Memories.Add(Memory);

	// Trim to max
	if (Memories.Num() > MaxMemoriesPerNpc)
	{
		Memories.RemoveAt(0, Memories.Num() - MaxMemoriesPerNpc);
	}
}

bool UMemorySystem::GetDialogueContext(const FString& NpcId, const FString& PlayerId,
	const FString& PlayerName, const TMap<FString, float>& EmotionalState,
	int32 MaxMemories, FSoulDialogueContext& OutContext)
{
	const FEntityRelation* RelPtr = Relations.Find(PlayerId);
	if (!RelPtr) return false;

	OutContext.NpcId = NpcId;
	OutContext.PlayerId = PlayerId;
	OutContext.PlayerName = PlayerName.IsEmpty() ? RelPtr->EntityName : PlayerName;
	OutContext.Relation = *RelPtr;
	OutContext.EmotionalState = EmotionalState;

	// Get recent memories about this player
	TArray<FEntityMemory> PlayerMemories;
	for (const FEntityMemory& Mem : Memories)
	{
		if (Mem.EntityId == PlayerId)
			PlayerMemories.Add(Mem);
	}
	if (PlayerMemories.Num() > MaxMemories)
	{
		PlayerMemories.RemoveAt(0, PlayerMemories.Num() - MaxMemories);
	}
	OutContext.RecentMemories = PlayerMemories;

	return true;
}

TArray<FEntityMemory> UMemorySystem::GetEntityMemories(const FString& EntityId, int32 MaxCount)
{
	TArray<FEntityMemory> Result;
	for (int32 i = Memories.Num() - 1; i >= 0 && Result.Num() < MaxCount; --i)
	{
		if (Memories[i].EntityId == EntityId)
			Result.Add(Memories[i]);
	}
	return Result;
}

bool UMemorySystem::GetRelation(const FString& EntityId, FEntityRelation& OutRelation)
{
	const FEntityRelation* RelPtr = Relations.Find(EntityId);
	if (!RelPtr) return false;
	OutRelation = *RelPtr;
	return true;
}

TArray<FEntityRelation> UMemorySystem::GetAllRelations()
{
	TArray<FEntityRelation> Result;
	for (const auto& KV : Relations)
	{
		Result.Add(KV.Value);
	}
	return Result;
}

void UMemorySystem::DecayRelations(float DeltaTime)
{
	for (auto& KV : Relations)
	{
		FEntityRelation& Relation = KV.Value;
		float DecayFactor = 1.0f - (RelationDecayRate * DeltaTime * 0.1f);
		Relation.Strength *= DecayFactor;

		float TrustDecay = 1.0f - (RelationDecayRate * DeltaTime * 0.05f);
		Relation.Trust *= TrustDecay;
	}
}

void UMemorySystem::UpdateTime(float DeltaTime)
{
	GameTime += DeltaTime;
}

void UMemorySystem::Clear()
{
	Relations.Empty();
	Memories.Empty();
	MemoryCounter = 0;
}

FString UMemorySystem::GenerateMemoryDescription(ESoulEventType EventType, const FString& EntityName)
{
	switch (EventType)
	{
	case ESoulEventType::PlayerAttacked:    return FString::Printf(TEXT("被%s攻击"), *EntityName);
	case ESoulEventType::PlayerHelped:      return FString::Printf(TEXT("被%s帮助"), *EntityName);
	case ESoulEventType::ItemReceived:      return FString::Printf(TEXT("从%s处收到物品"), *EntityName);
	case ESoulEventType::ItemLost:          return FString::Printf(TEXT("被%s夺走物品"), *EntityName);
	case ESoulEventType::QuestCompleted:    return FString::Printf(TEXT("与%s完成任务"), *EntityName);
	case ESoulEventType::QuestFailed:       return FString::Printf(TEXT("与%s任务失败"), *EntityName);
	case ESoulEventType::NpcInsulted:       return FString::Printf(TEXT("被%s侮辱"), *EntityName);
	case ESoulEventType::NpcComplimented:   return FString::Printf(TEXT("被%s称赞"), *EntityName);
	case ESoulEventType::GiftGiven:         return FString::Printf(TEXT("收到%s的礼物"), *EntityName);
	case ESoulEventType::EnvironmentChanged:return TEXT("环境发生变化");
	case ESoulEventType::RelationshipChanged:return FString::Printf(TEXT("与%s关系改变"), *EntityName);
	case ESoulEventType::CombatStarted:     return FString::Printf(TEXT("与%s开始战斗"), *EntityName);
	case ESoulEventType::CombatEnded:       return FString::Printf(TEXT("与%s结束战斗"), *EntityName);
	case ESoulEventType::DeathWitnessed:    return FString::Printf(TEXT("目睹%s死亡"), *EntityName);
	case ESoulEventType::TimePassed:        return TEXT("时间流逝");
	default:                                return FString::Printf(TEXT("与%s发生未知事件"), *EntityName);
	}
}

File: unity/NeshamaSDK/Runtime/SoulEngine/Memory/MemorySystem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Neshama.SoulEngine.Emotion;

namespace Neshama.SoulEngine.Memory
{
    /// <summary>
    /// NPC relation with an entity.
    /// Ported from Python npc_memory_bridge.py EntityRelation.
    /// </summary>
    [Serializable]
    public class EntityRelation
    {
        public string entityId;
        public string entityName;
        public string relationType = "neutral";
        public float strength = 0.3f;     // -1 to 1
        public float trust = 0.3f;        // 0 to 1
        public float lastInteractionTime; // game time
        public int interactionCount;

        public string GetStrengthCategory()
        {
            if (strength >= 0.8f) return "intimate";
            if (strength >= 0.6f) return "close";
            if (strength >= 0.4f) return "friendly";
            if (strength >= 0.2f) return "neutral";
            if (strength >= 0.0f) return "distant";
            return "hostile";
        }
    }

    /// <summary>
    /// Dialogue context for NPC-Player interaction.
    /// Ported from Python npc_memory_bridge.py DialogueContext.
    /// </summary>
    public class DialogueContext
    {
        public string npcId;
        public string playerId;
        public string playerName;
        public EntityRelation relation;
        public List<EntityMemory> recentMemories = new List<EntityMemory>();
        public Dictionary<string, float> emotionalState = new Dictionary<string, float>();

        /// <summary>
        /// Generate prompt parts for LLM dialogue generation.
        /// Ported from Python DialogueContext.to_prompt_parts().
        /// </summary>
        public List<string> ToPromptParts(int maxMemories = 3)
        {
            var parts = new List<string>();

            if (relation != null)
            {
                parts.Add($"你与{playerName}的关系是{relation.relationType}，" +
                          $"强度{relation.strength:F2}，信任度{relation.trust:F2}");
            }

            if (recentMemories != null && recentMemories.Count > 0)
            {
                var memDescs = recentMemories
                    .Take(maxMemories)
                    .Select(m => $"{m.eventType}({Truncate(m.description, 20)}...)");
                parts.Add($"你最近与{playerName}的交互：{string.Join("、", memDescs)}");
            }

            if (emotionalState != null)
            {
                var significant = emotionalState
                    .Where(kv => kv.Value > 0.2f)
                    .OrderByDescending(kv => kv.Value)
                    .Select(kv => $"{kv.Key}({kv.Value:F2})");
                var sigList = significant.ToList();
                if (sigList.Count > 0)
                {
                    parts.Add($"你当前情绪：{string.Join("、", sigList)}");
                }
            }

            return parts;
        }

        private static string Truncate(string s, int maxLen)
        {
            if (s == null) return "";
            return s.Length <= maxLen ? s : s.Substring(0, maxLen);
        }
    }

    /// <summary>
    /// Memory system for an NPC. Manages entity relations and memories.
    /// Ported from Python npc_memory_bridge.py NPCMemoryBridge (per-NPC portion).
    /// </summary>
    public class MemorySystem
    {
        /// <summary>Maximum memories per NPC before oldest are dropped.</summary>
        public int maxMemoriesPerNpc = 50;

        /// <summary>Relation decay rate per second.</summary>
        public float relationDecayRate = 0.001f;

        // Entity relations: entityId → EntityRelation
        private Dictionary<string, EntityRelation> _relations = new Dictionary<string, EntityRelation>();

        // Memories stored as list (ordered by time)
        private List<EntityMemory> _memories = new List<EntityMemory>();

        // Memory ID counter
        private int _memoryCounter = 0;

        // Game time tracker
        private float _gameTime = 0f;

        /// <summary>
        /// Current game time (in seconds). Updated by Tick().
        /// </summary>
        public float GameTime => _gameTime;

        // ── Game Event Processing ────────────────────────────────────────────────

        /// <summary>
        /// Event-to-relation mapping rules.
        /// Ported from Python EVENT_RELATION_MAPPINGS.
        /// </summary>
        private static readonly Dictionary<GameEventType, RelationMapping> EventRelationMappings =
            new Dictionary<GameEventType, RelationMapping>
        {
            { GameEventType.PlayerAttacked, new RelationMapping { relation = "hostile", strengthDelta = 0.3f, trustDelta = -0.2f } },
            { GameEventType.PlayerHelped, new RelationMapping { relation = "ally", strengthDelta = 0.3f, trustDelta = 0.2f } },
            { GameEventType.ItemReceived, new RelationMapping { relation = "friendly", strengthDelta = 0.1f, trustDelta = 0.1f } },
            { GameEventType.ItemLost, new RelationMapping { relation = "neutral", strengthDelta = -0.1f, trustDelta = -0.1f } },
            { GameEventType.QuestCompleted, new RelationMapping { relation = "ally", strengthDelta = 0.3f, trustDelta = 0.3f } },
            { GameEventType.QuestFailed, new RelationMapping { relation = "disappointed", strengthDelta = -0.2f, trustDelta = -0.1f } },
            { GameEventType.NpcInsulted, new RelationMapping { relation = "hostile", strengthDelta = 0.2f, trustDelta = -0.3f } },
            { GameEventType.NpcComplimented, new RelationMapping { relation = "friendly", strengthDelta = 0.2f, trustDelta = 0.2f } },
            { GameEventType.GiftGiven, new RelationMapping { relation = "friendly", strengthDelta = 0.2f, trustDelta = 0.3f } },
            { GameEventType.EnvironmentChanged, new RelationMapping { relation = "aware", strengthDelta = 0.1f, trustDelta = 0f } },
            { GameEventType.RelationshipChanged, new RelationMapping { relation = "connected", strengthDelta = 0.2f, trustDelta = 0f } },
            { GameEventType.CombatStarted, new RelationMapping { relation = "hostile", strengthDelta = 0.2f, trustDelta = 0f } },
            { GameEventType.CombatEnded, new RelationMapping { relation = "tense", strengthDelta = 0.0f, trustDelta = 0f } },
            { GameEventType.DeathWitnessed, new RelationMapping { relation = "shaken", strengthDelta = 0.2f, trustDelta = 0f } },
            { GameEventType.TimePassed, new RelationMapping { relation = "neutral", strengthDelta = -0.05f, trustDelta = 0f } },
        };

        // ── Core Methods ─────────────────────────────────────────────────────────

        /// <summary>
        /// Process a game event and update entity relation + memory.
        /// Ported from Python NPCMemoryBridge.on_game_event().
        /// </summary>
        public void OnGameEvent(GameEventType eventType, float intensity, string entityId, string entityName,
            Dictionary<string, float> emotionalContext = null)
        {
            RelationMapping mapping;
            if (!EventRelationMappings.TryGetValue(eventType, out mapping))
                return;

            // Get or create relation
            EntityRelation relation;
            if (!_relations.TryGetValue(entityId, out relation))
            {
                relation = new EntityRelation
                {
                    entityId = entityId,
                    entityName = entityName,
                    strength = Math.Max(0f, Math.Min(1f, 0.3f + mapping.strengthDelta * intensity)),
                    trust = Math.Max(0f, Math.Min(1f, 0.3f + mapping.trustDelta * intensity)),
                    relationType = mapping.relation,
                    lastInteractionTime = _gameTime,
                    interactionCount = 1,
                };
                _relations[entityId] = relation;
            }
            else
            {
                relation.strength = Math.Max(-1f, Math.Min(1f, relation.strength + mapping.strengthDelta * intensity));
                relation.trust = Math.Max(0f, Math.Min(1f, relation.trust + mapping.trustDelta * intensity));
                relation.relationType = mapping.relation;
                relation.lastInteractionTime = _gameTime;
                relation.interactionCount++;
            }

            // Create memory entry
            _memoryCounter++;
            var memory = new EntityMemory
            {
                memoryId = $"mem_{_memoryCounter}",
                entityId = entityId,
                entityName = entityName,
                eventType = eventType.ToString(),
                description = GenerateMemoryDescription(eventType, entityName),
                timestamp = _gameTime,
                emotionalContext = emotionalContext != null
                    ? new Dictionary<string, float>(emotionalContext)
                    : new Dictionary<string, float>(),
                trustAtTime = relation.trust,
            };
            _memories.Add(memory);

            // Trim to max
            if (_memories.Count > maxMemoriesPerNpc)
            {
                _memories = _memories.GetRange(_memories.Count - maxMemoriesPerNpc, maxMemoriesPerNpc);
            }
        }

        /// <summary>
        /// Get dialogue context for NPC-Player interaction.
        /// </summary>
        public DialogueContext GetDialogueContext(
            string npcId, string playerId, string playerName = null,
            Dictionary<string, float> emotionalState = null, int maxMemories = 5)
        {
            EntityRelation relation;
            if (!_relations.TryGetValue(playerId, out relation))
                return null;

            var memories = _memories
                .Where(m => m.entityId == playerId)
                .ToList();
            if (memories.Count > maxMemories)
                memories = memories.GetRange(memories.Count - maxMemories, maxMemories);

            return new DialogueContext
            {
                npcId = npcId,
                playerId = playerId,
                playerName = playerName ?? relation.entityName,
                relation = relation,
                recentMemories = memories,
                emotionalState = emotionalState ?? new Dictionary<string, float>(),
            };
        }

        /// <summary>
        /// Get memories about a specific entity.
        /// </summary>
        public List<EntityMemory> GetEntityMemories(string entityId, int maxCount = 10)
        {
            var result = _memories
                .Where(m => m.entityId == entityId)
                .ToList();
            if (result.Count > maxCount)
                result = result.GetRange(result.Count - maxCount, maxCount);
            result.Reverse(); // Most recent first
            return result;
        }

        /// <summary>
        /// Get NPC's relation with an entity.
        /// </summary>
        public EntityRelation GetRelation(string entityId)
        {
            EntityRelation rel;
            return _relations.TryGetValue(entityId, out rel) ? rel : null;
        }

        /// <summary>
        /// Get all relations for this NPC.
        /// </summary>
        public List<EntityRelation> GetAllRelations()
        {
            return _relations.Values.ToList();
        }

        /// <summary>
        /// Decay relations over time. Call in Tick().
        /// Ported from Python NPCMemoryBridge.decay_relations().
        /// </summary>
        public void DecayRelations(float deltaTime)
        {
            foreach (var relation in _relations.Values)
            {
                float decayFactor = 1f - (relationDecayRate * deltaTime * 0.1f);
                relation.strength *= decayFactor;

                float trustDecay = 1f - (relationDecayRate * deltaTime * 0.05f);
                relation.trust *= trustDecay;
            }
        }

        /// <summary>
        /// Update game time. Call in Tick().
        /// </summary>
        public void UpdateTime(float deltaTime)
        {
            _gameTime += deltaTime;
        }

        /// <summary>
        /// Clear all data.
        /// </summary>
        public void Clear()
        {
            _relations.Clear();
            _memories.Clear();
            _memoryCounter = 0;
        }

        // ── Memory Description Generation ────────────────────────────────────────

        private static string GenerateMemoryDescription(GameEventType eventType, string entityName)
        {
            switch (eventType)
            {
                case GameEventType.PlayerAttacked: return $"被{entityName}攻击";
                case GameEventType.PlayerHelped: return $"被{entityName}帮助";
                case GameEventType.ItemReceived: return $"从{entityName}处收到物品";
                case GameEventType.ItemLost: return $"被{entityName}夺走物品";
                case GameEventType.QuestCompleted: return $"与{entityName}完成任务";
                case GameEventType.QuestFailed: return $"与{entityName}任务失败";
                case GameEventType.NpcInsulted: return $"被{entityName}侮辱";
                case GameEventType.NpcComplimented: return $"被{entityName}称赞";
                case GameEventType.GiftGiven: return $"收到{entityName}的礼物";
                case GameEventType.EnvironmentChanged: return "环境发生变化";
                case GameEventType.RelationshipChanged: return $"与{entityName}关系改变";
                case GameEventType.CombatStarted: return $"与{entityName}开始战斗";
                case GameEventType.CombatEnded: return $"与{entityName}结束战斗";
                case GameEventType.DeathWitnessed: return $"目睹{entityName}死亡";
                case GameEventType.TimePassed: return "时间流逝";
                default: return $"与{entityName}发生未知事件";
            }
        }

        // ── Helper Struct ────────────────────────────────────────────────────────

        private struct RelationMapping
        {
            public string relation;
            public float strengthDelta;
            public float trustDelta;
        }
    }
}
